api/utils/sshutils/ssh.go

```diff
@@ -250,3 +260,263 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
+	fileComment := "teleport-generated-ppk"
+
+	// create a buffer to hold the elements needed to generate the MAC
+	macInput := new(bytes.Buffer)
+
+	// string: the algorithm-name header field.
+	macKeyType, err := getRFC4251String([]byte(keyType))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.LittleEndian, macKeyType)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macKeyType to buffer: %T", err))
+	}
+
+	// string: the encryption-type header field.
+	macEncryptionType, err := getRFC4251String([]byte(encryptionType))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macEncryptionType)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macEncryptionType to buffer: %T", err))
+	}
+
+	// string: the key-comment-string header field.
+	macComment, err := getRFC4251String([]byte(fileComment))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macComment)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macComment to buffer: %T", err))
+	}
+
+	// base64-decode the Teleport public key, as we need its binary representation to generate the MAC
+	teleportPublicKeyDecoded := make([]byte, base64.StdEncoding.EncodedLen(len(teleportPublicKeyBytes)))
+	teleportPublicKeyByteCount, err := base64.StdEncoding.Decode(teleportPublicKeyDecoded, teleportPublicKeyBytes)
+	publicKeyData := make([]byte, teleportPublicKeyByteCount)
+	copy(publicKeyData, teleportPublicKeyDecoded)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("could not base64-decode public key: %T, got %v bytes successfully", err, teleportPublicKeyByteCount))
+	}
+	// append the decoded public key bytes to the MAC buffer
+	macPublicKeyData, err := getRFC4251String(publicKeyData)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macPublicKeyData)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macPublicKeyData to buffer: %T", err))
+	}
+
+	// append our PPK-formatted private key bytes to the MAC buffer
+	macPrivateKeyData, err := getRFC4251String(ppkPrivateKey.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(macInput, binary.BigEndian, macPrivateKeyData)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macPrivateKeyData to buffer: %T", err))
+	}
+
+	// as per the PPK spec, the key for the MAC is blank when the PPK file is unencrypted.
+	// therefore, the key is a zero-length byte slice.
+	hmacHash := hmac.New(sha256.New, []byte{})
+	// generate the MAC using HMAC-SHA-256
+	hmacHash.Write(macInput.Bytes())
+	macString := hex.EncodeToString(hmacHash.Sum(nil))
+
+	// build the string-formatted output PPK file
+	var ppk strings.Builder
+	ppk.WriteString(fmt.Sprintf("PuTTY-User-Key-File-3: %v\n", keyType))
```

Comment:
`strings.Builder` implements io.Writer, so there's no need to generate a string first with `Sprintf` and then write to the buffer. Just write this instead:

```
fmt.Fprintf(ppk, "PuTTY-User-Key-File-3: %v\n", keyType)
```

Also, the same goes for `bytes.Buffer`, and it looks like what you ultimately need at the end is a `[]byte`, not a string, so I would replace the `strings.Builder` with a `*bytes.Buffer`